src/linker/Linker/DynamicDependency.cs

```diff
@@ -86,40 +86,28 @@ public DynamicDependency (DynamicallyAccessedMemberTypes memberTypes, string typ
 			if (args.Count < 1 || args.Count > 3)
 				return null;
 
-			// First argument is string or DynamicallyAccessedMemberTypes
-			string? memberSignature = args[0].Value as string;
-			if (args.Count == 1)
-				return memberSignature == null ? null : new DynamicDependency (memberSignature);
-			DynamicallyAccessedMemberTypes? memberTypes = null;
-			if (memberSignature == null) {
-				var argType = args[0].Type;
-				if (!argType.IsTypeOf<DynamicallyAccessedMemberTypes> ())
-					return null;
-				try {
-					memberTypes = (DynamicallyAccessedMemberTypes) args[0].Value;
-				} catch (InvalidCastException) { }
-				if (memberTypes == null)
-					return null;
-			}
-
-			// Second argument is Type for ctors with two args, string for ctors with three args
-			if (args.Count == 2) {
-				if (!(args[1].Value is TypeReference type))
-					return null;
-				return memberSignature == null ? new DynamicDependency (memberTypes!.Value, type) : new DynamicDependency (memberSignature, type);
-			}
-			Debug.Assert (args.Count == 3);
-			if (!(args[1].Value is string typeName))
-				return null;
-
-			// Third argument is assembly name
-			if (!(args[2].Value is string assemblyName))
-				return null;
-
-			var dynamicDependency = memberSignature == null ? new DynamicDependency (memberTypes!.Value, typeName, assemblyName) : new DynamicDependency (memberSignature, typeName, assemblyName);
-			dynamicDependency.OriginalAttribute = ca;
-
-			return dynamicDependency;
+			DynamicDependency? result = args[0].Value switch
+			{
+				string stringMemberSignature => args.Count switch
+				{
+					1 => new DynamicDependency (stringMemberSignature),
+					2 when args[1].Value is TypeReference type => new DynamicDependency (stringMemberSignature, type),
+					3 when args[1].Value is string typeName && args[2].Value is string assemblyName => new DynamicDependency (stringMemberSignature, typeName, assemblyName),
+					_ => null,
+				},
+				int memberTypes => args.Count switch
```

Comment:
This changes the semantics of the code - not sure if it matters though. Previous we would check that the type of the `args[0].Type` was `DynamicallyAccessedMemberTypes`, now we simply look at the value and use it if it's an int. So this can probably consume more cases without failing.

Additionally before we casted the value to the enum and if it failed we also "failed", now we don't catch that exception. So if the cast fails for some reason, this would throw. That said, I'm actually not sure if casting int to enum can every fail.

Reply:
I'd prefer to keep the stricter behavior with the `IsTypeOf` check, but it makes little difference in practice.

The cast check was there to prevent using an enum backed by the wrong type. It would fail if the user tried to redefine the attribute to take `enum DynamicallyAccessedMemberTypes : UInt32` for example. I think hard-coding the implementation type `int` in the pattern match should cover the same cases.